src/contractions.jl

```diff
@@ -120,34 +115,18 @@ end
 
 
 function LinearAlgebra.dot(O::AbstractMPO, ψ::AbstractMPS)
-    S = promote_type(eltype(ψ), eltype(O))
-    T = typeof(ψ)
-    ϕ = T.name.wrapper(S, length(ψ))
-    for (i, (A, B)) ∈ enumerate(zip(O, ψ))
+    MPS([
         @matmul N[(x, a), σ, (y, b)] := sum(η) A[x, σ, y, η] * B[a, η, b]
-        ϕ[i] = N
-    end
-    ϕ
-end
-
-
-function dot!(ψ::AbstractMPS, O::AbstractMPO)
-    for (i, (A, B)) ∈ enumerate(zip(ψ, O))
-        @matmul N[(x, a), σ, (y, b)] := sum(η) B[x, σ, y, η] * A[a, η, b]
-        ψ[i] = N
-    end
+        for (A, B) ∈ zip(O, ψ)
+    ])
 end
 
 
 function LinearAlgebra.dot(O1::AbstractMPO, O2::AbstractMPO)
-    S = promote_type(eltype(O1), eltype(O2))
-    T = typeof(O1)
-    O = T.name.wrapper(S, length(O1))
-    for (i, (A, B)) ∈ enumerate(zip(O1, O2))
+    MPO([
```

Comment:
This will fail with `CuMPO`, please restore previous approach.

Reply:
But this should be faster - less copy operations if I'm not mistaken.